pallets/order-book/src/lib.rs

```diff
@@ -721,123 +730,80 @@ pub mod pallet {
-			order_id: T::OrderIdNonce,
-			buy_amount: T::Balance,
-			sell_rate_limit: T::SellRatio,
-			min_fulfillment_amount: T::Balance,
-			validate: impl FnOnce(&OrderOf<T>) -> DispatchResult,
+			mut order: OrderOf<T>,
+			amount_out: T::Balance,
+			price: OrderPrice<T::SellRatio>,
+			min_amount_out: T::Balance,
 		) -> DispatchResult {
-			let max_sell_amount = <Orders<T>>::try_mutate_exists(
-				order_id,
-				|maybe_order| -> Result<T::Balance, DispatchError> {
-					let order = maybe_order.as_mut().ok_or(Error::<T>::OrderNotFound)?;
+			let min_fulfillment_amount_out = T::DecimalConverter::to_asset_balance(
+				T::MinFulfillmentAmountNative::get(),
+				order.asset_out_id,
+			)?;
+
+			Self::validate_amount(amount_out, min_fulfillment_amount_out, min_amount_out)?;
 
-					let max_sell_amount = Self::convert_with_ratio(
-						order.asset_in_id,
+			// ensure proper amount can be, and is reserved of outgoing currency for updated
+			// order.
+			// Also minimise reserve/unreserve operations.
+			if amount_out != order.amount_out || price != order.price {
+				if amount_out > order.amount_out {
+					let sell_reserve_diff = amount_out.ensure_sub(order.amount_out)?;
+					T::TradeableAsset::hold(
+						order.asset_out_id,
+						&(),
+						&order.placing_account,
+						sell_reserve_diff,
+					)?;
+				} else {
+					let sell_reserve_diff = order.amount_out.ensure_sub(amount_out)?;
+					T::TradeableAsset::release(
 						order.asset_out_id,
-						sell_rate_limit,
-						buy_amount,
+						&(),
+						&order.placing_account,
+						sell_reserve_diff,
+						Precision::Exact,
 					)?;
+				}
+			};
+			order.amount_out = amount_out;
+			order.price = price;
+			order.min_fulfillment_amount_out = min_fulfillment_amount_out;
 
-					// ensure proper amount can be, and is reserved of outgoing currency for updated
-					// order.
-					// Also minimise reserve/unreserve operations.
-					if buy_amount != order.buy_amount || sell_rate_limit != order.max_sell_rate {
-						if max_sell_amount > order.max_sell_amount {
-							let sell_reserve_diff =
-								max_sell_amount.ensure_sub(order.max_sell_amount)?;
-							T::TradeableAsset::hold(
-								order.asset_out_id,
-								&(),
-								&account,
-								sell_reserve_diff,
-							)?;
-						} else {
-							let sell_reserve_diff =
-								order.max_sell_amount.ensure_sub(max_sell_amount)?;
-							T::TradeableAsset::release(
-								order.asset_out_id,
-								&(),
-								&account,
-								sell_reserve_diff,
-								Precision::Exact,
-							)?;
-						}
-					};
-					order.buy_amount = buy_amount;
-					order.max_sell_rate = sell_rate_limit;
-					order.min_fulfillment_amount = min_fulfillment_amount;
-					order.max_sell_amount = max_sell_amount;
-
-					validate(order)?;
-
-					Ok(max_sell_amount)
-				},
-			)?;
+			Orders::<T>::insert(order.order_id, order.clone());
+			UserOrders::<T>::insert(order.placing_account, order.order_id, order.clone());
```

Comment:
NIT: I feel it is much cleaner and less boilerplate if inserted twice instead of mutating both storage

Reply:
This way, we are dropping the existence check of the order which is otherwise performed by `try_mutate_exists`.

Reply:
The existence is already checked when the order is previously retrived, in `Orders::<T>::get()`.